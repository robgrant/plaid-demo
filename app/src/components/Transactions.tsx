import { gql, useQuery } from "@apollo/client";
import {
  Table,
  TableBody,
  TableCell,
  TableRow,
  Typography,
} from "@material-ui/core";
import React from "react";
import {
  GetTransactions,
  GetTransactionsVariables,
} from "./__generated__/GetTransactions";

interface Props {
  accessToken: string;
  accountID: string;
}

export const query = gql`
  query GetTransactions($accessToken: String!, $accountID: ID!) {
    transactions(accessToken: $accessToken, accountID: $accountID) {
      id
      name
      date
      isoCurrencyCode
      amount
    }
  }
`;

export const Transactions: React.FunctionComponent<Props> = ({
  accessToken,
  accountID,
}) => {
  const { loading, error, data } = useQuery<
    GetTransactions,
    GetTransactionsVariables
  >(query, { variables: { accessToken, accountID } });

  if (loading) {
    return <Typography>Loading transaction data...</Typography>;
  }

  if (error || !data) {
    return <>{error?.message ?? "Error"}</>;
  }

  if (data.transactions.length === 0) {
    return <Typography>No transactions</Typography>;
  }

  return (
    <Table>
      <TableBody>
        {data.transactions.map((transaction) => {
          return (
            <TableRow key={transaction.id}>
              <TableCell>
                <Typography>{transaction.name}</Typography>
              </TableCell>
              <TableCell>
                <Typography>{transaction.date}</Typography>
              </TableCell>
              <TableCell>
                <Typography>
                  {new Intl.NumberFormat("en-US", {
                    style: "currency",
                    currency: transaction.isoCurrencyCode ?? "USD",
                  }).format(transaction.amount)}
                </Typography>
              </TableCell>
            </TableRow>
          );
        })}
      </TableBody>
    </Table>
  );
};
