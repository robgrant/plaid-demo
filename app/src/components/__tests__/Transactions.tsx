import React from "react";
import renderer from "react-test-renderer";
import { Transactions, query } from "../Transactions";
import { MockedProvider, MockedResponse } from "@apollo/client/testing";

describe("Transactions", () => {
  it("renders when transactions are returned", async () => {
    const mocks: MockedResponse[] = [
      {
        request: {
          query,
          variables: { accessToken: "accessToken", accountID: "accountID" },
        },
        result: {
          data: {
            transactions: [
              {
                id: "wP7AMWDnr4SpEaR9GlyetbBqZWGwkgurzG7Rb",
                name: "Uber 063015 SF**POOL**",
                date: "2020-10-12",
                isoCurrencyCode: "USD",
                amount: 5.4,
              },
              {
                id: "5lQxPvzKqWHwbvVzR8AqSd7NVyP36GiZXvPyD",
                name: "United Airlines",
                date: "2020-10-10",
                isoCurrencyCode: "USD",
                amount: -500,
              },
              {
                id: "JpravMdgo3sdeo6LD3lGfjgM9N7vlptdXArMr",
                name: "McDonald's",
                date: "2020-10-09",
                isoCurrencyCode: "USD",
                amount: 12,
              },
            ],
          },
        },
      },
    ];

    const component = renderer.create(
      <MockedProvider mocks={mocks} addTypename={false}>
        <Transactions accessToken="accessToken" accountID="accountID" />
      </MockedProvider>
    );

    expect(component.toJSON()).toMatchSnapshot();

    await renderer.act(() => new Promise((resolve) => setTimeout(resolve, 0)));

    expect(component.toJSON()).toMatchSnapshot();
  });

  it("renders when the account has no transactions", async () => {
    const mocks: MockedResponse[] = [
      {
        request: {
          query,
          variables: { accessToken: "accessToken", accountID: "accountID" },
        },
        result: {
          data: {
            transactions: [],
          },
        },
      },
    ];

    const component = renderer.create(
      <MockedProvider mocks={mocks} addTypename={false}>
        <Transactions accessToken="accessToken" accountID="accountID" />
      </MockedProvider>
    );

    expect(component.toJSON()).toMatchSnapshot();

    await renderer.act(() => new Promise((resolve) => setTimeout(resolve, 0)));

    expect(component.toJSON()).toMatchSnapshot();
  });
});
