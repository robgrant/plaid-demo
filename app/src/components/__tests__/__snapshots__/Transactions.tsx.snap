// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Transactions renders when the account has no transactions 1`] = `
<p
  className="MuiTypography-root MuiTypography-body1"
>
  Loading transaction data...
</p>
`;

exports[`Transactions renders when the account has no transactions 2`] = `
<p
  className="MuiTypography-root MuiTypography-body1"
>
  No transactions
</p>
`;

exports[`Transactions renders when transactions are returned 1`] = `
<p
  className="MuiTypography-root MuiTypography-body1"
>
  Loading transaction data...
</p>
`;

exports[`Transactions renders when transactions are returned 2`] = `
<table
  className="MuiTable-root"
  role={null}
>
  <tbody
    className="MuiTableBody-root"
    role={null}
  >
    <tr
      className="MuiTableRow-root"
      role={null}
    >
      <td
        aria-sort={null}
        className="MuiTableCell-root MuiTableCell-body"
      >
        <p
          className="MuiTypography-root MuiTypography-body1"
        >
          Uber 063015 SF**POOL**
        </p>
      </td>
      <td
        aria-sort={null}
        className="MuiTableCell-root MuiTableCell-body"
      >
        <p
          className="MuiTypography-root MuiTypography-body1"
        >
          2020-10-12
        </p>
      </td>
      <td
        aria-sort={null}
        className="MuiTableCell-root MuiTableCell-body"
      >
        <p
          className="MuiTypography-root MuiTypography-body1"
        >
          $5.40
        </p>
      </td>
    </tr>
    <tr
      className="MuiTableRow-root"
      role={null}
    >
      <td
        aria-sort={null}
        className="MuiTableCell-root MuiTableCell-body"
      >
        <p
          className="MuiTypography-root MuiTypography-body1"
        >
          United Airlines
        </p>
      </td>
      <td
        aria-sort={null}
        className="MuiTableCell-root MuiTableCell-body"
      >
        <p
          className="MuiTypography-root MuiTypography-body1"
        >
          2020-10-10
        </p>
      </td>
      <td
        aria-sort={null}
        className="MuiTableCell-root MuiTableCell-body"
      >
        <p
          className="MuiTypography-root MuiTypography-body1"
        >
          -$500.00
        </p>
      </td>
    </tr>
    <tr
      className="MuiTableRow-root"
      role={null}
    >
      <td
        aria-sort={null}
        className="MuiTableCell-root MuiTableCell-body"
      >
        <p
          className="MuiTypography-root MuiTypography-body1"
        >
          McDonald's
        </p>
      </td>
      <td
        aria-sort={null}
        className="MuiTableCell-root MuiTableCell-body"
      >
        <p
          className="MuiTypography-root MuiTypography-body1"
        >
          2020-10-09
        </p>
      </td>
      <td
        aria-sort={null}
        className="MuiTableCell-root MuiTableCell-body"
      >
        <p
          className="MuiTypography-root MuiTypography-body1"
        >
          $12.00
        </p>
      </td>
    </tr>
  </tbody>
</table>
`;
